### ISSUE TITLE
warnings/crashes in newer Lustre versions related to use of ioctls

### ISSUE THREAD

**Issue by shanedsnyder** [darshan-hpc/darshan#270]
In GitLab by @shanedsnyder on Apr 2, 2020, 11:25

Darshan's Lustre module uses ioctl to determine striping information for files, but this ability has been deprecated in newer Lustre versions, leading to warnings like:

```
using old ioctl(LL_IOC_LOV_GETSTRIPE) on [0x2c002048c:0x25:0x0], use llapi_layout_get_by_path()
```

Another user has reported crashes if the Lustre module is not disabled, using Lustre version 2.11.0.300_cray_102_g3dbace1.

We should look at using the suggested API for retrieving this info, and confirm its overhead.

**Comment by shanedsnyder** [comment]
In GitLab by @glennklockwood on Apr 2, 2020, 18:36

Is this a superset of #260?

Either way, Andreas Dilger suggested the following way back when we were creating the Lustre module:

> The newer `llapi_layout_get_by_fd()` routine is probably the best choice, if it is available in your system, since it is abstracting the layout internals since they will be changing in the future. It is using `fgetxattr(fd, "lustre.lov", buf, buflen)` internally, which is also as efficient as the `ioctl()` call, but also works on compute nodes that do not have the `ioctl()` plumbing (e.g. BlueGene).  In general, the `llapi_layout_*()` APIs are the way forward. 

The `llapi_layout_get_by_path()` is what we started with, but it does not scale because it requires full path resolution on every client for a fd that Darshan already has.  Hopefully the suggested `llapi_layout_get_by_fd()` still exists and works.

**Comment by shanedsnyder** [comment]
In GitLab by @carns on Apr 8, 2020, 14:13

Thanks for the pointer/reminder @glennklockwood .  FYI that API call is present in headers on Theta right now:

```
[carns@thetalogin4 lustre]$ pwd
/usr/include/lustre
[carns@thetalogin4 lustre]$ grep layout_get_by_fd *
lustreapi.h:struct llapi_layout *llapi_layout_get_by_fd(int fd, uint32_t flags);
```

There are really three sub-tasks in here:
* make Darshan use that API (or similar) if present
* test performance of said API
* regardless of any of the above, test the crash being triggered via ioctl.  We need to confirm what's going on here and report upstream if needed

**Comment by shanedsnyder** [comment]
In GitLab by @roblatham00 on May 21, 2020, 11:51

I tried out https://xgitlab.cels.anl.gov/darshan/darshan/blob/master/darshan-test/2.x/llapi-perf.c on a virtual machine with kernel `3.10.0-957.27.2.el7_lustre.x86_64` and lustre's pre-built `lustre-2.13.53_207_g1c54733` rpms.

I do see the warning about a deprecated api in dmesg, but no crashes.

**Comment by shanedsnyder** [comment]
In GitLab by @carns on May 25, 2020, 20:23

Possible that it has been fixed?

FYI, here are the Jira issues on Lustre related to the ioctl that we suspect to be (or have been) problematic:

https://jira.whamcloud.com/browse/LU-12580?jql=text%20~%20%22LL_IOC_LOV_GETSTRIPE%2C%22

I have not dug through them, but the first one mentions Darshan.

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Mar 18, 2021, 00:43

mentioned in merge request !89

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Mar 18, 2021, 00:47

closed via commit 332f4ebbed94d64fd90a0c82b236fa8978035193

**Comment by shanedsnyder** [comment]
In GitLab by @shanedsnyder on Mar 29, 2021, 19:56

mentioned in commit 332f4ebbed94d64fd90a0c82b236fa8978035193
